songpal/device.py

```diff
@@ -449,38 +454,35 @@ def clear_notification_callbacks(self):
         """Clear all notification callbacks."""
         self.callbacks.clear()
 
-    async def listen_notifications(self, fallback_callback=None):
+    async def listen_notifications(
+        self, fallback_callback: Optional[NotificationCallback] = None
+    ):
         """Listen for notifications from the device forever.
 
         Use :func:on_notification: to register what notifications to listen to.
         """
         tasks = []
 
-        async def handle_notification(notification):
-            if type(notification) not in self.callbacks:
-                if not fallback_callback:
-                    _LOGGER.debug("No callbacks for %s", notification)
-                    # _LOGGER.debug("Existing callbacks for: %s" % self.callbacks)
-                else:
-                    await fallback_callback(notification)
-                return
-            for cb in self.callbacks[type(notification)]:
-                await cb(notification)
-
-        for serv in self.services.values():
-            tasks.append(
-                asyncio.ensure_future(
-                    serv.listen_all_notifications(handle_notification)
+        async def handle_notification(notification: ChangeNotification) -> None:
+            callbacks = self.callbacks.get(type(notification), [fallback_callback])
+            if not callbacks:
+                _LOGGER.debug("No callbacks defined for %s", notification)
+            else:
```

Comment:
If I'm not misreading this, would it make sense to return when no callbacks are set, and avoid indent for the else-block altogether?

Reply:
I have no preference between the two, done :)

Reply:
Thanks! Less nesting makes code cleaner, and clean code is good ;-)